API for gix-archive

svetli-n opened the pull request:
----

Ok for [Byron](https://github.com/Byron) review the PR on video?

- [x] I give my permission to record review and upload on YouTube publicly

If I think the review will be helpful for the community, then I might record and publish a video.


Related to https://github.com/Byron/gitoxide/issues/783

svetli-n commented:
@Byron Here is a very high level API, assuming that command line options and .gitattributes related processing are done elsewhere (eg gix crate)

Byron commented on gix-archive/src/archive.rs:
```diff
@@ -0,0 +1,44 @@
+
+/// Iterates over `_entries` (ready to be added to an archive files) and writes to `_destination` according to archive `opts`
+pub fn archive<R: Read, W: Write>(
+    _entries: Iter<'_, File<R>>,
```
Doing it like this means the data cannot be streamed, and everything has to be in memory in advance, thus doesn't scale.

svetli-n replied:
I thought that since iter is lazy, only a single file has to be in mem, or not?

Byron replied:
It's hard-coded to be a slice iterator, which implies all of it is in memory in advance. What's seems to be intended is `impl IntoIterator<Item = File<R>>`. I see issues happening when trying to fill in `R` with an actual type and recommend using a `Vec<u8>` instead - I consider huge file support optional as `gitoxide` wouldn't be good with huge files as well. Ideally there is a way to reuse these data `Vec`s as well, but I think once things became clearer it will get there naturally.

svetli-n replied:
ok, changed to `impl IntoIterator<File>`(File has Vec<u8>, not Read)

svetli-n replied:
> I see issues happening when trying to fill in `R` with an actual type

What issues are you seeing?

Byron commented on gix-archive/src/archive.rs:
```diff
@@ -0,0 +1,44 @@
+
+/// Available formats
+#[derive(PartialEq)]
+pub enum Format {
```
I think with a scaling implementation and support for more customization there will be a trait of sorts to be implement by each archiving backend. However, I can also imagine that the selection of these backends could be hidden and an `enum` allows to choose them easily (along with feature toggles to pull in their dependencies on demand).

svetli-n replied:
Added `Custom(FromConfig)` variant

Byron replied:
I am biased towards using a trait for this, but in theory all that can be done differently too and I am open to that. This is exploration so ideally there is tests for the anticipated usecases.

Byron reviewed (changes requested):
Thanks for giving it a try!

There are three major requirements that I don't think are covered in this sample yet:

* scalability - deal with arbitrarily sized repos (while assuming individual files that fit into memory easily)
* attributes are handled in crate - these are complicated along with special attributes that only `git archive` cares about. This makes it impossible to delegate all attribute logic to users.
* support for custom backends isn't possible with a preset `enum`, but there should be a way to select standard archive types out of the box as well.

I am saying all of this knowing that you might have all that in mind, but since there is no PR description I wouldn't know what's planned or what's a known limitation.

svetli-n commented:
> * attributes are handled in crate - these are complicated along with special attributes that only `git archive` cares about. This makes it impossible to delegate all attribute logic to users.

i don't fully understand this. My understanding is that attributes handling, being more general than archive should happen outside of archive crate. My be there will be some cache that can be queried with a path, built from hierarchical attributes reading (from multiple locations). The archive crate would be able to query the cache for a path and if it has ignore attribute, it will skip adding it to the archive. If there is no ignore attribute and there is expand attribute, it will expand (e.g. will fillin commit ID) and then it will add the file to the archive. I think this is what git does, looking [here](https://github.com/git/git/blob/master/archive.c#L167) and [here](https://github.com/git/git/blob/master/archive.c#L187) 
What do you think?

svetli-n commented on gix-archive/src/lib.rs:
```diff
@@ -1,2 +1,4 @@
 #![deny(rust_2018_idioms)]
 #![forbid(unsafe_code)]
+
+pub mod archive;
```
Move to lib, but import path is still `gix_archive::archive::archive`, may be I misunderstood something..

Byron replied:
Modules that are private can be used for 'storage', and public `use` can export symbols from there to reduce the depth of the import path.

```rust
mod archive {
pub struct Foo;
}
pub use archive::Foo;
```

svetli-n replied:
Changed like that, thanks.

svetli-n commented on gix-archive/src/archive.rs:
```diff
@@ -0,0 +1,44 @@
+
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+pub enum Error {
```
ok, changed to struct.

svetli-n replied:
are we happy with it so far?

Byron replied:
This should be:

```Rust
    #[derive(Debug, thiserror::Error)]
    #[allow(missing_docs)]
    #[error("{msg}")]
    pub struct FailedToArchiveError {
        msg: String,
    }
```

svetli-n replied:
Changed like that, thanks.

svetli-n commented:
> * scalability - deal with arbitrarily sized repos (while assuming individual files that fit into memory easily)

Now its `impl IntoIterator<File>`

Byron commented:
> > * attributes are handled in crate - these are complicated along with special attributes that only `git archive` cares about. This makes it impossible to delegate all attribute logic to users.
> 
> i don't fully understand this. My understanding is that attributes handling, being more general than archive should happen outside of archive crate. My be there will be some cache that can be queried with a path, built from hierarchical attributes reading (from multiple locations). The archive crate would be able to query the cache for a path and if it has ignore attribute, it will skip adding it to the archive. If there is no ignore attribute and there is expand attribute, it will expand (e.g. will fillin commit ID) and then it will add the file to the archive. I think this is what git does, looking [here](https://github.com/git/git/blob/master/archive.c#L167) and [here](https://github.com/git/git/blob/master/archive.c#L187) What do you think?

Thanks for the code references, very helpful! Yes, the `gix-archive` code needs access to git attributes in some way, and access to a way to apply filters that would also be applied during worktree checkouts. This system doesn't exist yet, I ought to build it. On top of that, `gix-archive` has its own special kind of filter that does the substitution, that seems to be running after the working tree conversion.

svetli-n commented:
> This system doesn't exist yet, I ought to build it.

Once there, I'll add it to the API's input. Is there something else we are missing in terms of API? Or I will focus on implementation details, such as using tar (and other) crate(s) meanwhile

Byron commented:
> Once there, I'll add it to the API's input. Is there something else we are missing in terms of API? Or I will focus on implementation details, such as using tar (and other) crate(s) meanwhile

I think this can be boiled down to applying `tree-to-worktree-filters` outside of this crate, and thinking about it, maybe this should just be documented as we are at some point receiving the data which should be the data that would go to the worktree. Then there should be a way to also obtain git attributes of a path (maybe along with the data) so we can look at them to see if the crate-internal filters need to be applied. `gix-attributes` can be pulled in to get these data types.

And with that in mind, it should be possible to continue to sketch parts of the API like you suggested.

Byron reviewed (commented):
> Once there, I'll add it to the API's input. Is there something else we are missing in terms of API? Or I will focus on implementation details, such as using tar (and other) crate(s) meanwhile

I think this can be boiled down to applying `tree-to-worktree-filters` outside of this crate, and thinking about it, maybe this should just be documented as we are at some point receiving the data which should be the data that would go to the worktree. Then there should be a way to also obtain git attributes of a path (maybe along with the data) so we can look at them to see if the crate-internal filters need to be applied. `gix-attributes` can be pulled in to get these data types.

And with that in mind, it should be possible to continue to sketch parts of the API like you suggested.

svetli-n commented:
> > Once there, I'll add it to the API's input. Is there something else we are missing in terms of API? Or I will focus on implementation details, such as using tar (and other) crate(s) meanwhile
> 
> I think this can be boiled down to applying `tree-to-worktree-filters` outside of this crate, and thinking about it, maybe this should just be documented as we are at some point receiving the data which should be the data that would go to the worktree. Then there should be a way to also obtain git attributes of a path (maybe along with the data) so we can look at them to see if the crate-internal filters need to be applied. `gix-attributes` can be pulled in to get these data types.
> 
> And with that in mind, it should be possible to continue to sketch parts of the API like you suggested.

Great!

`export-subs` expands all placeholders, documented [here](https://git-scm.com/docs/git-log#Documentation/git-log.txt-emnem), such as `%C(…​)` (color specification) or `%H` (commit hash). We need a parser that handles it. There is no such parser or am I missing it? If I would to write it, is there a similar parser I can use as reference?

Byron reviewed (commented):
> export-subs expands all placeholders, documented [here](https://git-scm.com/docs/git-log#Documentation/git-log.txt-emnem), such as %C(…​) (color specification) or %H (commit hash). We need a parser that handles it. There is no such parser or am I missing it? If I would to write it, is there a similar parser I can use as reference?

Oh my, pandoras box, again 😅. It looks like first there is a wrapper https://github.com/git/git/blob/140b9478dad5d19543c1cb4fd293ccec228f1240/archive.c#L41 and then it applies the entire formatting logic of `git log` which seems to provide access to all information in a commit in various formats.

This capability of substituting `git log` style formats into a buffer (or writing a new with substitutions applied) seems useful even outside of the realm of `gix-archive` - maybe a `gix-log` crate is in order to handle the substitution for that.

In any case, it seems like this is something that should be added to the tracking issue and I'd say it can be skipped for the MVP as long as `gix-archive` stops if such an unsupported attribute is applied to a file. What do you think?

Further, when looking at the API, without the git-specific substitution capabilities this crate really is a standard archiving utility that doesn't know or care about it git at all, it's quite interesting to see it unwrap like that.

svetli-n commented on gix-archive/src/lib.rs:
```diff
@@ -1,2 +1,63 @@
+    }
+}
+
+pub use archive::{archive, File, Format, Options};
```
How about `gix_archive::write`?

svetli-n commented:
> In any case, it seems like this is something that should be added to the tracking issue and I'd say it can be skipped for the MVP as long as `gix-archive` stops if such an unsupported attribute is applied to a file. What do you think?

Yes, updated https://github.com/Byron/gitoxide/issues/783 and also moved acting on .gitattributes out of MVP since there is no way to query them by path yet, or do you think it will be available?

svetli-n commented:
> Oh my, pandoras box, again 😅. It looks like first there is a wrapper https://github.com/git/git/blob/140b9478dad5d19543c1cb4fd293ccec228f1240/archive.c#L41 and then it applies the entire formatting logic of `git log` which seems to provide access to all information in a commit in various formats.

Yep, I'll stop saying/expecting that a seemingly simple feature is indeed simple :-)

svetli-n commented:
> Further, when looking at the API, without the git-specific substitution capabilities this crate really is a standard archiving utility that doesn't know or care about it git at all, it's quite interesting to see it unwrap like that.

Yes, as it is now it depends on other crates 
- to get a list of files to be archived
- to query .gitattributes
- to expand placeholders

Is it valuable to have archive functionality without these or how common is it to not have to worry about them?

Byron commented:
> Is it valuable to have archive functionality without these or how common is it to not have to worry about them?

That's a very interesting question, thanks for bringing it up!

> to get a list of files to be archived

Right now the input is an object hash. From there it can decode objects (`gix-object`) and a tree (`gix-traverse`) to obtain paths and query their data with a `find` closure that is passed from the outside. For a moment I forgot that the input of paths and their data via iterator is just for *additional* files.

This also means it needs to have a way to access the filtering machinery for which there *will be* `gix-filter`, which is controlled by attributes which can probably be passed in via `gix_worktree::fs::Cache` or an equivalent closure to abstract that detail. After all, we really only care about a set of attributes per paths, partly interesting for us for our own substitution, and partially interesting to the one applying filters. That will probably also happen via a closure as `gix-config` also plays a role there, so `gix-filter` alone is probably not going to cut it. As long as there is a way to obtain attributes an external `data to worktree conversion` step can be added later as well.

> to expand placeholders

I think the `git archive` part of the placeholders needs to be implemented for the MVP to be able to reject files that would need substitution (while it is not implemented via something like `gix-log` or maybe `gix-substitute` :).

All in all, I think the crate will do as much as it can to deliver an archive of what *would* be checked out into the worktree, even though some parts of it it simply can't do yet or needs outside help with as it can't own filtering capabilities, nor can it own attributes.

For me, this is coming together nicely even though for now it's just a theory.

svetli-n commented:
> Right now the input is an object hash. From there it can decode objects (`gix-object`) and a tree (`gix-traverse`) to obtain paths and query their data with a `find` closure that is passed from the outside. For a moment I forgot that the input of paths and their data via iterator is just for _additional_ files.

I meant the iterator to be for _all_ files. We get them similar to [here](https://github.com/Byron/gitoxide/blob/main/gix/examples/stats.rs#L40) in `core::repository::archive` and pass them to archive crate or better with the closure?

Byron commented:
> I meant the iterator to be for _all_ files. We get them similar to [here](https://github.com/Byron/gitoxide/blob/main/gix/examples/stats.rs#L40) in `core::repository::archive` (to be added) and pass them to archive crate or better with the closure?

The rightful question came up of what the `gix-archive` crate has to do with git if it's (nearly) entirely decoupled. And since a useful plumbing crate should be as usable as possible when using it directly, the question should be ask if it can do more or encapsulate more knowledge about archiving git trees. That's where the insight of traversing the tree *here* comes from, along with the realization that the iterator interface should only be used for adding additional files.

> In order to know which files need substitution we need to query .gitattributes for export-subst attribute for a path via gix-log or maybe gix-substitute, which won't be present for the MVP or?

I expect this capability to be ready soon enough, even though it might be added a little later than the rest of it.

>  We could know which files need substitution by looking for `$Format:PLACEHOLDERS$` in a file but it would be temp fix.

I think that's a great idea - while there are no attributes to guide the substitution, we can look for placeholders in all files and simply assume - for the moment - that they should be substituted.

> By reject do you mean not adding them to the archive?

I think that can be explored - since it's a temporary measure, hard-failing seems to be the right thing. If somebody puts substitutions, they expect them to be applied *and* end up in the archive. If that can't be done, the archive can certainly be considered defective and that's what should be prevented with this measure.

svetli-n commented:
> the question should be ask if it can do more or encapsulate more knowledge about archiving git trees

Absolutely! I was not thinking about the crate being usable on its own.

svetli-n commented:
> hard-failing seems

It seems right to me as well

svetli-n commented:
> > the question should be ask if it can do more or encapsulate more knowledge about archiving git trees
> 
> Absolutely! I was not thinking about the crate being usable on its own. I'll add the closure to the API.

@Byron Added `Find` closure, like in `gix_traverse::tree::breadthfirst`. May be we have a final MVP API?

Byron commented on gix-archive/src/lib.rs:
```diff
@@ -1,2 +1,62 @@
+    pub enum Format {
+        Tar,
+        Zip,
+        TarGz,
```
Both of the compresed-`tar` formats can probably take a parameter for the compression level as part of the backend options.

Byron reviewed (commented):
Thanks! 

I have a few more comments but besides these, it should be OK to think about next steps. To me, this PR would be mergable if what was discussed here would be put into *rough* documentation so it becomes clearer how to do more intricate things like compressing with `zstd` or using an entirely different container format (i.e. decode `tar` stream and re-encode using [any] container).

Having these docs will help to stay aligned with it during implementation, while preserving all the knowledge that was built up so far even when the PR is merged.

svetli-n commented on gix-archive/src/lib.rs:
```diff
@@ -43,12 +37,17 @@ mod archive {
+    /// Finds files to be added to the archive my means of `_find` closure, optionally filtered by `_add_paths` and writes to `_destination` according to archive `opts`
+    pub fn write<W, Find>(
+        mut _find: Find,
+        _add_paths: Option<IntoIter<PathBuf>>,
```
> Lastly, having only a `PathBuf` doesn't cover the `--add-virtual-file=<path>:<content>` usecase which is the most general one, I wouldn't want to force API users to write temporary files.

Ok, but I mean `_add_paths` to be for https://git-scm.com/docs/git-archive#Documentation/git-archive.txt-ltpathgt not https://git-scm.com/docs/git-archive#Documentation/git-archive.txt---add-fileltfilegt or https://git-scm.com/docs/git-archive#Documentation/git-archive.txt---add-virtual-fileltpathgtltcontentgt, both not part of MVP, or shall we add them to the MVP?

Byron replied:
Then it would be best to remove the parameter entirely.

svetli-n commented on gix-archive/src/lib.rs:
```diff
@@ -1,2 +1,62 @@
+    }
+
+    /// Archive options, such as format, and later prefix, work-tree, etc
+    pub struct Options {
```
> After taking another look at possible options, it looks like there will be nothing else. Thus `Options` can be removed and `Format` is all there is left. Everything related to attributes and filters will be so essential that it can't be defaulted or be passed as options.

Other options are --worktree-attributes, --mtime, --prefix

svetli-n commented:
Added  ` _object_id: &oid,` param

svetli-n commented:
> put into _rough_ documentation

Are you thinking about doc strings in the code?

svetli-n commented on gix-archive/src/lib.rs:
```diff
@@ -1,2 +1,62 @@
+        Zip,
+        TarGz,
+        Tgz,
+        Custom(FromConfig),
```
> With this setup, the user won't be able to use other `Zip` like formats that would need calling an API to feed it data

Data comes primarily from the git tree or how do you mean?

Byron replied:
I was exploring how API users could use entirely different container formats. TAR and ZIP are the only supported ones right now and I think that's sufficient as `TAR` can easily be decoded to be placed into any other container.

Thus, please remove `Custom`. `TarGz` and `Tgz` are the same format and I wonder whether it's worth keeping `tar` compression as a convenience - after all that's easy done by providing a `zip` stream as `write`.

svetli-n replied:
Yes, Custom is gone. `TarGz` and `Tgz` follow git's options, which are probably familiar to users of the API. While the user can do anything in `std::io::Write` if they don't zip, we can do that for them -- friendlier API IMO -- or WDYT?

Byron commented:
> > put into _rough_ documentation
> 
> Sure! Are you thinking about doc strings in the code?

Yes, that's it.

svetli-n commented:
> Yes, that's it.

Done. It was all related to .gitattributes if I'm not missing something. Please take a look.
